lib/_http_agent.js

```diff
@@ -206,36 +206,42 @@ Agent.prototype.createSocket = function createSocket(req, options, cb) {
     }
     self.sockets[name].push(s);
     debug('sockets', name, self.sockets[name].length);
-
-    function onFree() {
-      self.emit('free', s, options);
-    }
-    s.on('free', onFree);
-
-    function onClose(err) {
-      debug('CLIENT socket onClose');
-      // This is the only place where sockets get removed from the Agent.
-      // If you want to remove a socket from the pool, just close it.
-      // All socket errors end in a close event anyway.
-      self.removeSocket(s, options);
-    }
-    s.on('close', onClose);
-
-    function onRemove() {
-      // We need this function for cases like HTTP 'upgrade'
-      // (defined by WebSockets) where we need to remove a socket from the
-      // pool because it'll be locked up indefinitely
-      debug('CLIENT socket onRemove');
-      self.removeSocket(s, options);
-      s.removeListener('close', onClose);
-      s.removeListener('free', onFree);
-      s.removeListener('agentRemove', onRemove);
-    }
-    s.on('agentRemove', onRemove);
+    self._installListeners(s,options);
     cb(null, s);
   }
 };
 
+Agent.prototype._installListeners = function installListeners(s, options) {
```

Comment:
Can you do this without attaching to the prototype. In its current form, this will become unofficial API that we'll have to maintain.